### DISCUSSION TITLE
How to restore cursor style on exit?

### DISCUSSION THREAD

**Discussion by lost22git** [sxyazi/yazi#1612] [1. Q&A]
My terminal cursor style is line, and it cannot restore it after quitting yazi (yazi curosr style is block)

**Comment by doorbuster** [comment]
I've noticed that happens in konsole.  When I quit yazi, the cursor is block, even though it was line before yazi.  This doesn't happen in kitty.  I have no solution, but a quick reset command brings it back.

**Reply by lost22git** [reply]
I do a fast check, it only happens in my windows terminal, and alacritty, wezterm works fine.

**Comment by sxyazi** [comment]
Yazi uses `DECSET` and `DECRQM` to request the terminal's cursor shape and blink status, so it can restore them when exiting:

https://github.com/sxyazi/yazi/blob/9483798459cbe8cb7dcca036a9463dc659fb0a41/yazi-fm/src/term.rs#L31-L33 

Unfortunately, some terminals don't support these requests, and a simple solution is to reset the cursor when exiting Yazi - if you're using the [shell wrapper](https://yazi-rs.github.io/docs/quick-start#shell-wrapper), you can add:

```diff
function y() {
	local tmp="$(mktemp -t "yazi-cwd.XXXXXX")"
	yazi "$@" --cwd-file="$tmp"
+	echo -e -n "\x1b[\x36 q" # change cursor to steady bar
	if cwd="$(cat -- "$tmp")" && [ -n "$cwd" ] && [ "$cwd" != "$PWD" ]; then
		builtin cd -- "$cwd"
	fi
	rm -f -- "$tmp"
}
```

into it to achieve this. See more available sequences at https://stackoverflow.com/questions/4416909/anyway-change-the-cursor-vertical-line-instead-of-a-box

**Reply by doorbuster** [reply]
Works great, thanks.  I hadn't noticed that the shell wrapper code (or at least the command) had changed.

**Comment by YHYJ** [comment]
Modify the function to allow run `yazi`, not `y`
```zsh
function yazi() {
    local tmp="$(mktemp -t "yazi-cwd.XXXXXX")"
    /usr/bin/yazi "$@" --cwd-file="$tmp"
    echo -e -n "\x1b[\x36 q" # change cursor to steady bar
    if cwd="$(cat -- "$tmp")" && [ -n "$cwd" ] && [ "$cwd" != "$PWD" ]; then
        builtin cd -- "$cwd"
    fi
    rm -f -- "$tmp"
}
```

**Comment by lost22git** [comment]
For powershell users, paste the below funcion to your `$PROFILE` file.

```powershell
enum CursorStyle
{
  block_blink = 1
  block = 2
  underline_blink = 3
  underline = 4
  bar_blink = 5
  bar = 6
}

<#
  .SYNOPSIS
  set cursor style
#>
function Set-CursorStyle
{
  param (
    # style
    [Parameter(Mandatory)]
    [CursorStyle]$Style
  )
  
  Write-Output "`e[$([int]$Style) q"
}
```

and you can set your cursor style after exiting yazi.

```powershell
Set-CursorStyle -Style bar
```

**Reply by shamone03** [reply]
Thanks, this worked. But to just reset the cursor style (if you had it configured in windows terminal for example) add another enum option that's equal to 0

**Comment by genieyr** [comment]
The same problem - cursor as block in editor.
My environment Windows Terminal, WSL, Archlinux, nano.

**Comment by bayang** [comment]
I had the same issue on Konsole, changing the shell integration this way fixed it for me : 

```
unction y() {
        local tmp="$(mktemp -t "yazi-cwd.XXXXXX")" cwd
        yazi "$@" --cwd-file="$tmp"
        if cwd="$(command cat -- "$tmp")" && [ -n "$cwd" ] && [ "$cwd" != "$PWD" ]; then
                builtin cd -- "$cwd"
        fi
        rm -f -- "$tmp"
  printf '\e[0 q\n'
}

```

Notice I added `  printf '\e[0 q\n'` at the end to revert cursor shape.
